libs/vkd3d/memory.c

```diff
@@ -1159,19 +1178,41 @@ HRESULT vkd3d_memory_allocator_flush_clears(struct vkd3d_memory_allocator *alloc
 }
 
 #define VKD3D_MEMORY_CLEAR_QUEUE_MAX_PENDING_BYTES (256ull << 20) /* 256 MiB */
+#define VKD3D_MEMORY_CLEAR_MAX_CPU_BYTES (4ull << 20) /* 4 MiB */
 
 static void vkd3d_memory_allocator_clear_allocation(struct vkd3d_memory_allocator *allocator,
         struct d3d12_device *device, struct vkd3d_memory_allocation *allocation)
 {
     struct vkd3d_memory_clear_queue *clear_queue = &allocator->clear_queue;
+    bool clear_with_gpu = false;
+    bool wait_for_clear = false;
 
     if (allocation->cpu_address)
     {
-        /* Probably faster than doing this on the GPU
-         * and having to worry about synchronization */
-        memset(allocation->cpu_address, 0, allocation->resource.size);
+        if (allocation->resource.size <= VKD3D_MEMORY_CLEAR_MAX_CPU_BYTES)
+        {
+            /* Probably faster than doing this on the GPU */
+            memset(allocation->cpu_address, 0, allocation->resource.size);
+        }
+        else
+        {
+            /*
+             * Do it on the GPU. No need to flush memory ranges since this is a fresh allocation. From Vulkan
+             * specification 11.2.12:
+             * Unmapping non-coherent memory does not implicitly flush the host mapped memory, and host writes that have
+             * not been flushed may not ever be visible to the device. However, implementations must ensure that writes
+             * that have not been flushed do not become visible to any other memory.
+             */
+            clear_with_gpu = true;
```

Comment:
This is quite questionable. Immediately stalling on the CPU is very bad. Why would potentially stalling for 10ms+ be preferable to a memset? How much data are we talking about here? How often? This is also clearing host memory, which means the clear is going to be PCI-e bound, which is far slower than DDR4 bandwidth.

Reply:
This could be HVV to be fair but even then it shouldn't be so slow, unless something is wrong with the memset implementation.

Reply:
Right, HVV could be an explanation ... Would be surprised if memset is that slow though. memset implementations like to use insane tricks for perf.